Future interface for ATen/Parallel

ilia-cher opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #21582 Parallelize bi-LSTM implementation
* **#21764 Future interface for ATen/Parallel**

Summary:
Adding std::future interface into ATen/Parallel for convenience

Test:
PARALLEL_BACKEND=NATIVE USE_CUDA=0 BLAS=MKL USE_MKLDNN=1 BUILD_BINARY=1 python setup.py develop --cmake
./build/bin/test_parallel
PARALLEL_BACKEND=OPENMP USE_CUDA=0 BLAS=MKL USE_MKLDNN=1 BUILD_BINARY=1 python setup.py develop --cmake
./build/bin/test_parallel
PARALLEL_BACKEND=NATIVE_TBB USE_TBB=1 USE_CUDA=0 BLAS=MKL USE_MKLDNN=1 BUILD_BINARY=1 python setup.py develop --cmake
./build/bin/test_parallel

Differential Revision: [D15816658](https://our.internmc.facebook.com/intern/diff/D15816658)

Commit:
Future interface for ATen/Parallel

Summary:
Adding std::future interface into ATen/Parallel for convenience

Commit:
Update on "Future interface for ATen/Parallel"

Future interface for ATen/Parallel

Summary:
Adding std::future interface into ATen/Parallel for convenience

gh-metadata: pytorch pytorch 21764 gh/ilia-cher/28/head

Commit:
Update on "Future interface for ATen/Parallel"

Future interface for ATen/Parallel

Summary:
Adding std::future interface into ATen/Parallel for convenience

gh-metadata: pytorch pytorch 21764 gh/ilia-cher/28/head

Commit:
Update on "Future interface for ATen/Parallel"

Future interface for ATen/Parallel

Summary:
Adding std::future interface into ATen/Parallel for convenience

gh-metadata: pytorch pytorch 21764 gh/ilia-cher/28/head

Commit:
Update on "Future interface for ATen/Parallel"

Future interface for ATen/Parallel

Summary:
Adding std::future interface into ATen/Parallel for convenience

gh-metadata: pytorch pytorch 21764 gh/ilia-cher/28/head

ilia-cher commented on aten/src/ATen/ParallelNative.cpp:
```diff
@@ -116,14 +116,29 @@ bool in_parallel_region() {
+}
 
+std::future<void> intraop_launch_future(std::function<void()> func) {
+  auto func_promise = std::make_shared<std::promise<void>>();
```
std::function requires the functor to be copyable, so, unfortunately, we cannot just move std::promise into lambda

Commit:
Update on "Future interface for ATen/Parallel"

Future interface for ATen/Parallel

Summary:
Adding std::future interface into ATen/Parallel for convenience

gh-metadata: pytorch pytorch 21764 gh/ilia-cher/28/head

Commit:
Update on "Future interface for ATen/Parallel"

Future interface for ATen/Parallel

Summary:
Adding std::future interface into ATen/Parallel for convenience

gh-metadata: pytorch pytorch 21764 gh/ilia-cher/28/head

cpuhrsch commented on aten/src/ATen/ParallelNative.h:
```diff
@@ -133,28 +132,27 @@ inline scalar_t parallel_reduce(
       internal::_unset_thread_num();
     };
 
-    std::vector<std::shared_ptr<ivalue::Future>> futures(num_tasks);
```
What's going to happen to "ivalue::Future"?

ilia-cher replied:
we still use it in JIT interpreter for fork/join implementation

Commit:
Update on "Future interface for ATen/Parallel"

Future interface for ATen/Parallel

Summary:
Adding std::future interface into ATen/Parallel for convenience

gh-metadata: pytorch pytorch 21764 gh/ilia-cher/28/head

cpuhrsch commented on aten/src/ATen/Parallel.h:
```diff
@@ -1,6 +1,8 @@
 #pragma once
 #include <ATen/ATen.h>
 
+#include <future>
```
I don't know how heavy of an include this is. If Parallel.h is supposed to be used widely it might make sense to move this somewhere else.

cpuhrsch replied:
Maybe create a Future.h?

ilia-cher replied:
discussed offline, will follow up with a more comprehensive reorg of parallel code files

Commit:
Update on "Future interface for ATen/Parallel"

Future interface for ATen/Parallel

Summary:
Adding std::future interface into ATen/Parallel for convenience

gh-metadata: pytorch pytorch 21764 gh/ilia-cher/28/head

Commit:
Update on "Future interface for ATen/Parallel"

Future interface for ATen/Parallel

Summary:
Adding std::future interface into ATen/Parallel for convenience

gh-metadata: pytorch pytorch 21764 gh/ilia-cher/28/head

Commit:
Update on "Future interface for ATen/Parallel"

Future interface for ATen/Parallel

Summary:
Adding std::future interface into ATen/Parallel for convenience

gh-metadata: pytorch pytorch 21764 gh/ilia-cher/28/head

Commit:
Update on "Future interface for ATen/Parallel"

Future interface for ATen/Parallel

Summary:
Adding std::future interface into ATen/Parallel for convenience

gh-metadata: pytorch pytorch 21764 gh/ilia-cher/28/head

Commit:
Update on "Future interface for ATen/Parallel"

Future interface for ATen/Parallel

Summary:
Adding std::future interface into ATen/Parallel for convenience

gh-metadata: pytorch pytorch 21764 gh/ilia-cher/28/head

ilia-cher commented:
using ivalue::Future
